https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

commit id: a17956e1c93b8ba51f844c5d1190009b3e4ea055

In [16]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [17]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
display(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
data.info()

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [18]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data[['Size nm^3']]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, train_size=0.9, random_state=42)
print(X_train)


     Temperature °C  Mols KCL
716              95       499
351             243       384
936              47       211
256             495       497
635              82       421
..              ...       ...
106             439        40
270             665       245
860             471       620
435             829       259
102             636       388

[900 rows x 2 columns]


## Part 3. Perform a Linear Regression

In [19]:
# Use sklearn to train a model on the training set

model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model

sample_dp = pd.DataFrame({'Temperature °C': [663], 'Mols KCL': [295]})
sample_pred = model.predict(sample_dp)

print("sample prediction =", sample_pred)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

train_score = model.score(X_train, y_train)
print("score of training set =", train_score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

coeffs = model.coef_
inter = model.intercept_

print("coefficients =", coeffs)
print("intercept =", inter)


sample prediction = [[469508.63709549]]
score of training set = 0.8610082946223788
coefficients = [[ 866.14641337 1032.69506649]]
intercept = [-409391.47958341]


The score of 0.861 suggests that the model is a pretty good fit for the data considering it is on a scale of 0 to 1.0. This score means that the model explains $\approx$ 86% of the variation of the size of the slime based on the inputs for temperature and mols of KCL




Sample equation: $E = mc^2$

equation:
$h(x) = -409391$ + $4866.15 \cdot X_0$ + $1032.7 \cdot X_1$

## Part 4. Use Cross Validation

In [20]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

scores = cross_val_score(model, X_train, y_train, cv=10)
print("scores =", scores)

# Report on their finding and their significance

print("mean of scores =", np.mean(scores))
print("standard deviation =", np.std(scores))

scores = [0.85704823 0.86460272 0.8150455  0.82683317 0.8746413  0.90201818
 0.88217921 0.84568669 0.8626558  0.85200682]
mean of scores = 0.8582717614710754
standard deviation = 0.02422520066423225


Overall, each fold of the data has a very similar score and the model fits very well for each fold, as the lowest score of all of the iterations is 0.815. The mean of all the scores signifies that the average score is $\approx$ 0.85827 which is very close to the training score in part 3. 


## Part 5. Using Polynomial Regression

In [21]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

poly = PolynomialFeatures(2)
poly_X = poly.fit_transform(X_train)

poly_model = LinearRegression()
poly_model.fit(poly_X, y_train)

poly_score = poly_model.score(poly_X, y_train)
print("score of poly_training set =", poly_score)

poly_coef = poly_model.coef_
poly_inter = poly_model.intercept_

print("poly_coefficients =", poly_coef)
print("poly_intercepts =", poly_inter)

# Report on the metrics and output the resultant equation as you did in Part 3.

poly_cross_scores = cross_val_score(poly_model, poly_X, y_train, cv=10)
print("cross scores =", poly_cross_scores)
print("mean of scores =", np.mean(poly_cross_scores))
print("standard deviation of scores =", np.std(poly_cross_scores))


score of poly_training set = 1.0
poly_coefficients = [[ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
   2.00000000e+00  2.85714287e-02]]
poly_intercepts = [2.04771059e-05]
cross scores = [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
mean of scores = 1.0
standard deviation of scores = 0.0


equation: $h(x) = 2.0477 \cdot 10^{-5}$ + $12X_1$ - $1.271 \cdot 10^{−7}X_2$ + $1.2649 \cdot 10^{-11}X_\frac{2}{1}$ + $2X_1X_2$ + $0.028571X_\frac{2}{2}$ 

The linear regression using the augmented dataset results in a model that perfectly fits the dataset and predicts the values. This means that the model explains 100% of the variation in the size of the slime based on the inputs for temperature and mols of KCL. The score of the model is 1.0 and when doing a cross validation over 10 folds of the dataset, each score is 1.0.